File: aseprite/services/export.py
"""Typed, policy-enforcing adapter around the Aseprite CLI and Lua bridge."""

from __future__ import annotations
import logging
import json
import tempfile
from pathlib import Path
from typing import Literal, cast
from ..errors import AsepriteMCPError
from ..models import (
    AtlasResult,
    BatchExportItem,
    BatchExportJob,
    BatchExportResult,
    BitmapFontResult,
    BitmapGlyphInput,
    ContactSheetResult,
    FrameExportInput,
    FrameExportItem,
    FrameExportResult,
    RenderResult,
    SliceExtractionInput,
    SliceExtractionItem,
    SliceExtractionResult,
    SpriteSheetResult,
)
from ..paths import (
    RENDER_EXTENSIONS,
    SPRITE_DOCUMENT_EXTENSIONS,
    SPRITE_INPUT_EXTENSIONS,
    publish_file,
    sha256_file,
    temporary_sibling,
)
from .runtime import AsepriteRuntime, MAX_CANVAS_PIXELS

logger = logging.getLogger(__name__)


class ExportService(AsepriteRuntime):
    """Export Aseprite operations."""

    async def render(
        self,
        source_path: str,
        output_path: str,
        *,
        frame: int | None,
        tag: str | None,
        layers: list[str],
        scale: float,
        overwrite: bool,
    ) -> RenderResult:
        source = self.paths.existing_file(source_path, extensions=SPRITE_INPUT_EXTENSIONS)
        output = self.paths.output_file(
            output_path, extensions=RENDER_EXTENSIONS, overwrite=overwrite
        )
        output_format = cast(Literal["png", "gif"], output.suffix.lower().lstrip("."))
        if frame is not None and tag is not None:
            raise AsepriteMCPError("INVALID_SELECTOR", "frame and tag are mutually exclusive")
        if output_format == "png" and tag is not None:
            raise AsepriteMCPError("INVALID_SELECTOR", "PNG rendering does not accept a tag")
        if frame is not None and frame < 0:
            raise AsepriteMCPError("INVALID_SELECTOR", "frame must be zero or greater")
        if not 0.1 <= scale <= 16:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "scale must be between 0.1 and 16")

        temporary = temporary_sibling(output)
        logger.info(
            "Rendering sprite source=%s output=%s format=%s frame=%s tag=%s "
            "scale=%s layers=%d overwrite=%s",
            source,
            output,
            output_format,
            frame,
            tag,
            f"{scale:g}",
            len(layers),
            overwrite,
        )
        arguments = ["--batch", "--noinapp"]
        for layer in layers:
            arguments.extend(["--layer", layer])
        if tag is not None:
            arguments.extend(["--tag", tag])
        selected_frame = frame
        if output_format == "png" and selected_frame is None:
            selected_frame = 0
        if selected_frame is not None:
            arguments.extend(["--frame-range", f"{selected_frame},{selected_frame}"])
        arguments.append(self._process_path(source))
        if scale != 1:
            arguments.extend(["--scale", f"{scale:g}"])
        arguments.extend(["--save-as", self._process_path(temporary)])

        async with self._locked([source, output]):
            try:
                await self._run(arguments, operation="render")
                if not temporary.is_file():
                    raise AsepriteMCPError(
                        "ASEPRITE_FAILED", "Aseprite did not create the rendered file"
                    )
                publish_file(temporary, output)
            finally:
                temporary.unlink(missing_ok=True)
        render_result = RenderResult(
            output_path=str(output),
            byte_size=output.stat().st_size,
            sha256=sha256_file(output),
            format=output_format,
            frame=selected_frame,
            tag=tag,
            scale=scale,
        )
        logger.info(
            "Render completed output=%s bytes=%d", output, render_result.byte_size
        )
        return render_result

    async def export_sprite_sheet(
        self,
        source_path: str,
        image_output_path: str,
        data_output_path: str,
        *,
        layout: Literal["horizontal", "vertical", "rows", "columns", "packed"],
        tag: str | None,
        layers: list[str],
        trim: bool,
        extrude: bool,
        border_padding: int,
        shape_padding: int,
        inner_padding: int,
        overwrite: bool,
    ) -> SpriteSheetResult:
        source = self.paths.existing_file(source_path, extensions=SPRITE_INPUT_EXTENSIONS)
        image_output = self.paths.output_file(
            image_output_path, extensions=frozenset({".png"}), overwrite=overwrite
        )
        data_output = self.paths.output_file(
            data_output_path, extensions=frozenset({".json"}), overwrite=overwrite
        )
        if image_output == data_output or image_output in sources or data_output in sources:
            raise AsepriteMCPError(
                "PATH_NOT_ALLOWED", "Atlas outputs must differ from each other and every source"
            )
        if image_output == data_output:
            raise AsepriteMCPError("PATH_NOT_ALLOWED", "Image and data outputs must differ")
        for value in (border_padding, shape_padding, inner_padding):
            if not 0 <= value <= 1024:
                raise AsepriteMCPError("LIMIT_EXCEEDED", "padding must be between 0 and 1024")

        image_temporary = temporary_sibling(image_output)
        data_temporary = temporary_sibling(data_output)
        logger.info(
            "Exporting sprite sheet source=%s image=%s data=%s layout=%s tag=%s "
            "layers=%d overwrite=%s",
            source,
            image_output,
            data_output,
            layout,
            tag,
            len(layers),
            overwrite,
        )
        arguments = ["--batch", "--noinapp"]
        for layer in layers:
            arguments.extend(["--layer", layer])
        if tag is not None:
            arguments.extend(["--tag", tag])
        arguments.append(self._process_path(source))
        arguments.extend(["--sheet-type", layout])
        if trim:
            arguments.append("--trim")
        if extrude:
            arguments.append("--extrude")
        if border_padding:
            arguments.extend(["--border-padding", str(border_padding)])
        if shape_padding:
            arguments.extend(["--shape-padding", str(shape_padding)])
        if inner_padding:
            arguments.extend(["--inner-padding", str(inner_padding)])
        arguments.extend(
            [
                "--sheet",
                self._process_path(image_temporary),
                "--format",
                "json-array",
                "--data",
                self._process_path(data_temporary),
            ]
        )

        async with self._locked([source, image_output, data_output]):
            try:
                await self._run(arguments, operation="export_sprite_sheet")
                if not image_temporary.is_file() or not data_temporary.is_file():
                    raise AsepriteMCPError(
                        "ASEPRITE_FAILED", "Aseprite did not create both sprite-sheet outputs"
                    )
                try:
                    sheet_data = json.loads(data_temporary.read_text(encoding="utf-8-sig"))
                except json.JSONDecodeError as exc:
                    raise AsepriteMCPError(
                        "ASEPRITE_FAILED", "Aseprite produced invalid sprite-sheet JSON"
                    ) from exc
                frame_count = len(sheet_data.get("frames", []))
                publish_file(image_temporary, image_output)
                publish_file(data_temporary, data_output)
            finally:
                image_temporary.unlink(missing_ok=True)
                data_temporary.unlink(missing_ok=True)

        sheet_result = SpriteSheetResult(
            image=self._file_result(image_output),
            data=self._file_result(data_output),
            layout=layout,
            frame_count=frame_count,
        )
        logger.info(
            "Sprite-sheet export completed image=%s data=%s frames=%d",
            image_output,
            data_output,
            frame_count,
        )
        return sheet_result

    async def preview(
        self,
        source_path: str,
        *,
        mode: Literal["frame", "sheet"],
        frame: int,
        layout: Literal["horizontal", "vertical", "rows", "columns", "packed"],
        tag: str | None,
        layers: list[str],
        scale: float,
    ) -> bytes:
        source = self.paths.existing_file(source_path, extensions=SPRITE_INPUT_EXTENSIONS)
        if frame < 0:
            raise AsepriteMCPError("INVALID_SELECTOR", "frame must be zero or greater")
        if mode == "frame" and tag is not None:
            raise AsepriteMCPError("INVALID_SELECTOR", "frame previews do not accept a tag")
        if not 0.1 <= scale <= 16:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "scale must be between 0.1 and 16")
        logger.info(
            "Previewing sprite source=%s mode=%s frame=%d tag=%s layers=%d scale=%s",
            source, mode, frame, tag, len(layers), f"{scale:g}",
        )
        with tempfile.TemporaryDirectory(
            prefix="aseprite-preview-", dir=self.bridge_temp_root
        ) as temporary_directory:
            output = Path(temporary_directory) / "preview.png"
            arguments = ["--batch", "--noinapp"]
            for layer in layers:
                arguments.extend(["--layer", layer])
            if tag is not None:
                arguments.extend(["--tag", tag])
            arguments.append(self._process_path(source))
            if scale != 1:
                arguments.extend(["--scale", f"{scale:g}"])
            if mode == "frame":
                arguments.extend(
                    ["--frame-range", f"{frame},{frame}", "--save-as", self._process_path(output)]
                )
            else:
                arguments.extend(
                    ["--sheet-type", layout, "--sheet", self._process_path(output)]
                )
            async with self._locked([source]):
                await self._run(arguments, operation="preview")
            if not output.is_file():
                raise AsepriteMCPError("ASEPRITE_FAILED", "Aseprite did not create the preview")
            byte_size = output.stat().st_size
            if byte_size > self.settings.max_capture_bytes:
                raise AsepriteMCPError(
                    "LIMIT_EXCEEDED",
                    "Preview exceeds the inline response size limit; reduce scale or frame count",
                )
            data = output.read_bytes()
        logger.info("Preview completed source=%s bytes=%d", source, len(data))
        return data

    async def render_contact_sheet(
        self,
        source_path: str,
        output_path: str,
        *,
        columns: int,
        scale: int,
        overwrite: bool,
    ) -> ContactSheetResult:
        if not 1 <= columns <= 64 or not 1 <= scale <= 16:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "columns or scale exceeds the limit")
        source = self.paths.existing_file(source_path, extensions=SPRITE_INPUT_EXTENSIONS)
        output = self.paths.output_file(
            output_path, extensions=frozenset({".png"}), overwrite=overwrite
        )
        temporary = temporary_sibling(output)
        async with self._locked([source, output]):
            try:
                result = await self._bridge(
                    "render_contact_sheet",
                    {
                        "source_path": self._process_path(source),
                        "output_path": self._process_path(temporary),
                        "columns": columns,
                        "scale": scale,
                        "max_pixels": MAX_CANVAS_PIXELS,
                    },
                )
                if not temporary.is_file():
                    raise AsepriteMCPError(
                        "ASEPRITE_FAILED", "Aseprite did not create the contact sheet"
                    )
                publish_file(temporary, output)
            finally:
                temporary.unlink(missing_ok=True)
        return ContactSheetResult(
            **self._file_result(output).model_dump(),
            frame_count=int(result["frame_count"]),
            columns=int(result["columns"]),
            rows=int(result["rows"]),
            scale=scale,
        )

    async def pack_atlas(
        self,
        source_paths: list[str],
        image_output_path: str,
        data_output_path: str,
        *,
        width: int,
        height: int,
        trim: bool,
        extrude: bool,
        border_padding: int,
        shape_padding: int,
        inner_padding: int,
        overwrite: bool,
    ) -> AtlasResult:
        if not 1 <= len(source_paths) <= 64:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "source_paths must contain 1 to 64 files")
        if width is not None and not 1 <= width <= 16_384:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "atlas width must be between 1 and 16384")
        if height is not None and not 1 <= height <= 16_384:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "atlas height must be between 1 and 16384")
        if any(value < 0 or value > 1024 for value in (
            border_padding, shape_padding, inner_padding
        )):
            raise AsepriteMCPError("LIMIT_EXCEEDED", "atlas padding must be between 0 and 1024")
        sources = [
            self.paths.existing_file(path, extensions=SPRITE_INPUT_EXTENSIONS)
            for path in source_paths
        ]
        image_output = self.paths.output_file(
            image_output_path, extensions=frozenset({".png"}), overwrite=overwrite
        )
        data_output = self.paths.output_file(
            data_output_path, extensions=frozenset({".json"}), overwrite=overwrite
        )
        image_temp, data_temp = temporary_sibling(image_output), temporary_sibling(data_output)
        arguments = ["--batch", "--noinapp", *[self._process_path(path) for path in sources]]
        arguments.append("--sheet-pack")
        if width is not None:
            arguments.extend(["--sheet-width", str(width)])
        if height is not None:
            arguments.extend(["--sheet-height", str(height)])
        if trim:
            arguments.append("--trim")
        if extrude:
            arguments.append("--extrude")
        for flag, value in (
            ("--border-padding", border_padding),
            ("--shape-padding", shape_padding),
            ("--inner-padding", inner_padding),
        ):
            if value:
                arguments.extend([flag, str(value)])
        arguments.extend([
            "--sheet", self._process_path(image_temp), "--format", "json-array",
            "--data", self._process_path(data_temp),
        ])
        async with self._locked([*sources, image_output, data_output]):
            try:
                await self._run(arguments, operation="pack_atlas")
                if not image_temp.is_file() or not data_temp.is_file():
                    raise AsepriteMCPError(
                        "ASEPRITE_FAILED", "Aseprite did not create atlas outputs"
                    )
                metadata = json.loads(data_temp.read_text(encoding="utf-8-sig"))
                frame_count = len(metadata.get("frames", []))
                atlas_size = metadata.get("meta", {}).get("size", {})
                atlas_width = int(atlas_size.get("w", 0))
                atlas_height = int(atlas_size.get("h", 0))
                if (
                    atlas_width < 1
                    or atlas_height < 1
                    or atlas_width > 16_384
                    or atlas_height > 16_384
                    or atlas_width * atlas_height > MAX_CANVAS_PIXELS
                ):
                    raise AsepriteMCPError(
                        "LIMIT_EXCEEDED", "Generated atlas dimensions exceed the server limit"
                    )
                publish_file(image_temp, image_output)
                publish_file(data_temp, data_output)
            except json.JSONDecodeError as exc:
                raise AsepriteMCPError(
                    "ASEPRITE_FAILED", "Aseprite produced invalid atlas JSON"
                ) from exc
            finally:
                image_temp.unlink(missing_ok=True)
                data_temp.unlink(missing_ok=True)
        return AtlasResult(
            image=self._file_result(image_output), data=self._file_result(data_output),
            source_count=len(sources), frame_count=frame_count,
        )

    async def extract_slices(
        self, source_path: str, *, extractions: list[SliceExtractionInput], overwrite: bool,
    ) -> SliceExtractionResult:
        if not 1 <= len(extractions) <= 256:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "extractions must contain 1 to 256 items")
        source = self.paths.existing_file(source_path, extensions=SPRITE_INPUT_EXTENSIONS)
        outputs = [self.paths.output_file(
            item.output_path, extensions=frozenset({".png"}), overwrite=overwrite
        ) for item in extractions]
        if len(set(outputs)) != len(outputs):
            raise AsepriteMCPError("INVALID_INPUT", "slice output paths must be unique")
        if source in outputs:
            raise AsepriteMCPError("PATH_NOT_ALLOWED", "slice output cannot replace its source")
        temporaries = [temporary_sibling(path) for path in outputs]
        async with self._locked([source, *outputs]):
            try:
                result = await self._bridge("extract_slices", {
                    "source_path": self._process_path(source),
                    "extractions": [{"name": item.name, "frame": item.frame,
                        "output_path": self._process_path(temp)}
                        for item, temp in zip(extractions, temporaries, strict=True)],
                })
                if not all(path.is_file() for path in temporaries):
                    raise AsepriteMCPError("ASEPRITE_FAILED", "Aseprite did not create every slice")
                for temporary, output in zip(temporaries, outputs, strict=True):
                    publish_file(temporary, output)
            finally:
                for temporary in temporaries:
                    temporary.unlink(missing_ok=True)
        items = [SliceExtractionItem(
            name=extraction.name, frame=extraction.frame, file=self._file_result(output),
            bounds=result["items"][index]["bounds"], pivot=result["items"][index].get("pivot"),
        ) for index, (extraction, output) in enumerate(zip(extractions, outputs, strict=True))]
        return SliceExtractionResult(
            source_path=str(source), sha256=sha256_file(source), items=items
        )

    async def batch_export(self, jobs: list[BatchExportJob]) -> BatchExportResult:
        if not 1 <= len(jobs) <= 64:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "jobs must contain 1 to 64 items")
        items: list[BatchExportItem] = []
        for job in jobs:
            try:
                result = await self.export_sprite_sheet(
                    job.source_path, job.image_output_path, job.data_output_path,
                    layout=job.layout, tag=job.tag, layers=job.layers, trim=job.trim,
                    extrude=job.extrude, border_padding=job.border_padding,
                    shape_padding=job.shape_padding, inner_padding=job.inner_padding,
                    overwrite=job.overwrite,
                )
                items.append(BatchExportItem(source_path=job.source_path, ok=True, result=result))
            except AsepriteMCPError as exc:
                items.append(BatchExportItem(
                    source_path=job.source_path, ok=False,
                    error_code=exc.code, error_message=exc.message,
                ))
        succeeded = sum(item.ok for item in items)
        return BatchExportResult(succeeded=succeeded, failed=len(items)-succeeded, items=items)

    async def export_frames(
        self, source_path: str, *, exports: list[FrameExportInput], overwrite: bool,
    ) -> FrameExportResult:
        if not 1 <= len(exports) <= 256:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "exports must contain 1 to 256 items")
        source = self.paths.existing_file(source_path, extensions=SPRITE_INPUT_EXTENSIONS)
        outputs = [self.paths.output_file(
            item.output_path, extensions=frozenset({".png"}), overwrite=overwrite
        ) for item in exports]
        if len(set(outputs)) != len(outputs) or source in outputs:
            raise AsepriteMCPError(
                "PATH_NOT_ALLOWED", "frame outputs must be unique and differ from the source"
            )
        temporaries = [temporary_sibling(path) for path in outputs]
        async with self._locked([source, *outputs]):
            try:
                await self._bridge("export_frames", {
                    "source_path": self._process_path(source),
                    "exports": [{"frame": item.frame, "output_path": self._process_path(temp)}
                        for item, temp in zip(exports, temporaries, strict=True)],
                })
                if not all(path.is_file() for path in temporaries):
                    raise AsepriteMCPError("ASEPRITE_FAILED", "Aseprite did not export every frame")
                for temporary, output in zip(temporaries, outputs, strict=True):
                    publish_file(temporary, output)
            finally:
                for temporary in temporaries:
                    temporary.unlink(missing_ok=True)
        return FrameExportResult(
            source_path=str(source), sha256=sha256_file(source),
            items=[FrameExportItem(frame=item.frame, file=self._file_result(output))
                   for item, output in zip(exports, outputs, strict=True)],
        )

    async def preview_nine_slice(
        self, source_path: str, *, slice_name: str, frame: int, width: int, height: int,
    ) -> bytes:
        source = self.paths.existing_file(source_path, extensions=SPRITE_DOCUMENT_EXTENSIONS)
        with tempfile.TemporaryDirectory(
            prefix="aseprite-nine-slice-preview-", dir=self.bridge_temp_root
        ) as directory:
            output = Path(directory) / "preview.png"
            async with self._locked([source]):
                await self._bridge("preview_nine_slice", {
                    "source_path": self._process_path(source), "slice": slice_name,
                    "frame": frame, "width": width, "height": height,
                    "output_path": self._process_path(output),
                    "max_pixels": MAX_CANVAS_PIXELS,
                })
            if not output.is_file():
                raise AsepriteMCPError("ASEPRITE_FAILED", "nine-slice preview was not created")
            if output.stat().st_size > self.settings.max_capture_bytes:
                raise AsepriteMCPError("LIMIT_EXCEEDED", "nine-slice preview exceeds inline limit")
            return output.read_bytes()

    async def export_bitmap_font(
        self, source_path: str, image_output_path: str, data_output_path: str, *,
        glyphs: list[BitmapGlyphInput], font_name: str, line_height: int,
        columns: int, padding: int, overwrite: bool,
    ) -> BitmapFontResult:
        if not 1 <= len(glyphs) <= 512:
            raise AsepriteMCPError("LIMIT_EXCEEDED", "glyphs must contain 1 to 512 entries")
        codepoints = [glyph.codepoint for glyph in glyphs]
        if len(set(codepoints)) != len(codepoints):
            raise AsepriteMCPError("INVALID_INPUT", "glyph codepoints must be unique")
        if any(0xD800 <= codepoint <= 0xDFFF for codepoint in codepoints):
            raise AsepriteMCPError("INVALID_INPUT", "surrogate codepoints are not valid Unicode")
        source = self.paths.existing_file(source_path, extensions=SPRITE_INPUT_EXTENSIONS)
        image_output = self.paths.output_file(
            image_output_path, extensions=frozenset({".png"}), overwrite=overwrite
        )
        data_output = self.paths.output_file(
            data_output_path, extensions=frozenset({".json"}), overwrite=overwrite
        )
        if len({source, image_output, data_output}) != 3:
            raise AsepriteMCPError("PATH_NOT_ALLOWED", "font input and outputs must be distinct")
        image_temp, data_temp = temporary_sibling(image_output), temporary_sibling(data_output)
        async with self._locked([source, image_output, data_output]):
            try:
                await self._bridge("export_bitmap_font", {
                    "source_path": self._process_path(source),
                    "image_output_path": self._process_path(image_temp),
                    "data_output_path": self._process_path(data_temp),
                    "glyphs": [glyph.model_dump() for glyph in glyphs],
                    "font_name": font_name, "line_height": line_height,
                    "columns": columns, "padding": padding,
                    "max_pixels": MAX_CANVAS_PIXELS,
                })
                if not image_temp.is_file() or not data_temp.is_file():
                    raise AsepriteMCPError(
                        "ASEPRITE_FAILED", "bitmap-font outputs were not created"
                    )
                publish_file(image_temp, image_output)
                publish_file(data_temp, data_output)
            finally:
                image_temp.unlink(missing_ok=True)
                data_temp.unlink(missing_ok=True)
        return BitmapFontResult(
            image=self._file_result(image_output), data=self._file_result(data_output),
            glyph_count=len(glyphs), line_height=line_height,
        )

File: aseprite/scripts/bridge/export.lua
operations.render_contact_sheet = function(input)
  local sprite = open_sprite(input.source_path)
  local frame_count = #sprite.frames
  local columns = math.min(input.columns, frame_count)
  local rows = math.ceil(frame_count / columns)
  local cell_width = sprite.width * input.scale + 2
  local cell_height = sprite.height * input.scale + 9
  local sheet_width = columns * cell_width
  local sheet_height = rows * cell_height
  if sheet_width * sheet_height > input.max_pixels then
    sprite:close()
    fail("LIMIT_EXCEEDED", "contact sheet exceeds the pixel limit")
  end
  local sheet = Image(sheet_width, sheet_height, ColorMode.RGB)
  sheet:clear(Rectangle(0, 0, sheet_width, sheet_height), Color { r = 32, g = 35, b = 42 })
  local label_color = app.pixelColor.rgba(230, 230, 230, 255)
  for frame_index = 1, #sprite.frames do
    local zero_index = frame_index - 1
    local column = zero_index % columns
    local row = math.floor(zero_index / columns)
    local origin_x = column * cell_width + 1
    local origin_y = row * cell_height
    draw_number(sheet, zero_index, origin_x, origin_y + 1, label_color)
    local frame_image = Image(sprite.width, sprite.height, ColorMode.RGB)
    frame_image:clear()
    frame_image:drawSprite(sprite, frame_index, Point(0, 0))
    if input.scale ~= 1 then
      frame_image:resize(sprite.width * input.scale, sprite.height * input.scale)
    end
    sheet:drawImage(frame_image, Point(origin_x, origin_y + 7))
  end
  sheet:saveAs(input.output_path)
  sprite:close()
  return { frame_count = frame_count, columns = columns, rows = rows }
end

operations.extract_slices = function(input)
  local sprite = open_sprite(input.source_path)
  local items = {}
  for _, extraction in ipairs(input.extractions) do
    local frame = extraction.frame + 1
    if frame < 1 or frame > #sprite.frames then fail("INVALID_SELECTOR", "slice frame is outside the sprite") end
    local slice = find_slice(sprite, extraction.name)
    if slice == nil then fail("INVALID_SELECTOR", "slice was not found: " .. extraction.name) end
    app.frame = sprite.frames[frame]
    local bounds = slice.bounds
    if bounds.width < 1 or bounds.height < 1 or bounds.x < 0 or bounds.y < 0 or
       bounds.x + bounds.width > sprite.width or bounds.y + bounds.height > sprite.height then
      fail("INVALID_SELECTOR", "slice bounds are empty or outside the sprite")
    end
    local rendered = render_frame_image(sprite, frame)
    local image = Image(rendered, bounds)
    if sprite.colorMode == ColorMode.INDEXED then
      image:saveAs {filename=extraction.output_path, palette=sprite.palettes[1]}
    else
      image:saveAs(extraction.output_path)
    end
    local pivot = nil
    if slice.pivot ~= nil then pivot = {x=slice.pivot.x, y=slice.pivot.y} end
    table.insert(items, {name=extraction.name, frame=extraction.frame,
      bounds=rectangle_info(bounds), pivot=pivot})
  end
  sprite:close()
  return {items=items}
end

operations.export_frames = function(input)
  local sprite = open_sprite(input.source_path)
  local items = {}
  for _, extraction in ipairs(input.exports) do
    local frame = extraction.frame + 1
    if frame < 1 or frame > #sprite.frames then fail("INVALID_SELECTOR", "export frame is outside the sprite") end
    local image = render_frame_image(sprite, frame)
    if sprite.colorMode == ColorMode.INDEXED then
      image:saveAs {filename=extraction.output_path, palette=sprite.palettes[1]}
    else image:saveAs(extraction.output_path) end
    table.insert(items, {frame=extraction.frame})
  end
  sprite:close()
  return {items=items}
end

local function draw_scaled_region(target,source,sx,sy,sw,sh,dx,dy,dw,dh)
  if sw<=0 or sh<=0 or dw<=0 or dh<=0 then return end
  local part=Image(source,Rectangle(sx,sy,sw,sh));if part.width~=dw or part.height~=dh then part:resize(dw,dh) end
  target:drawImage(part,Point(dx,dy))
end

operations.preview_nine_slice = function(input)
  local sprite=open_sprite(input.source_path);local slice=find_slice(sprite,input.slice)
  if slice==nil then fail("INVALID_SELECTOR","slice was not found") end
  local frame=input.frame+1;if frame<1 or frame>#sprite.frames then fail("INVALID_SELECTOR","frame is outside the sprite") end
  app.frame=sprite.frames[frame];local bounds=slice.bounds;local center=slice.center
  if center==nil then fail("INVALID_SELECTOR","slice does not define a nine-slice center") end
  if bounds.width<1 or bounds.height<1 or bounds.x<0 or bounds.y<0 or
    bounds.x+bounds.width>sprite.width or bounds.y+bounds.height>sprite.height then
    fail("INVALID_SELECTOR","slice bounds are empty or outside the sprite")
  end
  local left,top=center.x,center.y;local right=bounds.width-center.x-center.width
  local bottom=bounds.height-center.y-center.height
  if input.width<left+right or input.height<top+bottom then
    fail("INVALID_INPUT","preview target is smaller than fixed nine-slice borders") end
  if input.width*input.height>input.max_pixels then fail("LIMIT_EXCEEDED","nine-slice preview exceeds pixel limit") end
  local rendered=render_frame_image(sprite,frame);local source=Image(rendered,bounds)
  local spec=ImageSpec{width=input.width,height=input.height,colorMode=sprite.colorMode,
    transparentColor=sprite.transparentColor};local target=Image(spec);target:clear()
  local middle_w=input.width-left-right;local middle_h=input.height-top-bottom
  local xs={0,left,left+center.width};local sw={left,center.width,right};local dx={0,left,left+middle_w};local dw={left,middle_w,right}
  local ys={0,top,top+center.height};local sh={top,center.height,bottom};local dy={0,top,top+middle_h};local dh={top,middle_h,bottom}
  for row=1,3 do for col=1,3 do draw_scaled_region(target,source,xs[col],ys[row],sw[col],sh[row],dx[col],dy[row],dw[col],dh[row]) end end
  if sprite.colorMode==ColorMode.INDEXED then target:saveAs{filename=input.output_path,palette=sprite.palettes[1]}
  else target:saveAs(input.output_path) end
  sprite:close();return {width=input.width,height=input.height}
end

operations.export_bitmap_font = function(input)
  local sprite=open_sprite(input.source_path);local max_width,max_height=0,0
  for _,glyph in ipairs(input.glyphs) do
    if glyph.frame>=#sprite.frames or glyph.bounds.x<0 or glyph.bounds.y<0 or
      glyph.bounds.x+glyph.bounds.width>sprite.width or
      glyph.bounds.y+glyph.bounds.height>sprite.height then
      fail("INVALID_SELECTOR","glyph frame or bounds are outside the sprite")
    end
    max_width=math.max(max_width,glyph.bounds.width);max_height=math.max(max_height,glyph.bounds.height)
  end
  local columns=math.min(input.columns,#input.glyphs);local rows=math.ceil(#input.glyphs/columns)
  local cell_width,cell_height=max_width+input.padding*2,max_height+input.padding*2
  local width,height=columns*cell_width,rows*cell_height
  if width*height>input.max_pixels then fail("LIMIT_EXCEEDED","bitmap-font atlas exceeds pixel limit") end
  local atlas=Image(width,height,sprite.colorMode);atlas:clear();local entries={}
  for index,glyph in ipairs(input.glyphs) do
    local frame_image=render_frame_image(sprite,glyph.frame+1)
    local image=Image(frame_image,Rectangle(glyph.bounds.x,glyph.bounds.y,
      glyph.bounds.width,glyph.bounds.height))
    local x=((index-1)%columns)*cell_width+input.padding
    local y=math.floor((index-1)/columns)*cell_height+input.padding
    atlas:drawImage(image,Point(x,y))
    table.insert(entries,{codepoint=glyph.codepoint,frame=glyph.frame,x=x,y=y,
      width=glyph.bounds.width,height=glyph.bounds.height,
      advance=glyph.advance or glyph.bounds.width,bearing_x=glyph.bearing_x,
      bearing_y=glyph.bearing_y})
  end
  if sprite.colorMode==ColorMode.INDEXED then
    atlas:saveAs{filename=input.image_output_path,palette=sprite.palettes[1]}
  else atlas:saveAs(input.image_output_path) end
  write_all(input.data_output_path,json.encode({font=input.font_name,line_height=input.line_height,
    atlas={width=width,height=height,columns=columns,padding=input.padding},glyphs=entries}))
  sprite:close();return {glyph_count=#input.glyphs,line_height=input.line_height}
end
